optuna/study/_multi_objective.py

```diff
@@ -127,102 +80,128 @@ def _fast_non_dominated_sort(
-    num_infeasible_trials = np.count_nonzero(is_infeasible)
-    if num_infeasible_trials > 0:
-        _, ranks = np.unique(penalty[is_infeasible], return_inverse=True)
-        ranks += 1
-        nondomination_rank[is_infeasible] += 1 + bottom_rank + ranks
-        bottom_rank += np.max(ranks)
-        n_below -= num_infeasible_trials
+    n_infeasible = np.count_nonzero(is_infeasible)
+    if n_infeasible > 0:
+        _, ranks_in_infeas = np.unique(penalty[is_infeasible], return_inverse=True)
+        nondomination_rank[is_infeasible] = ranks_in_infeas + top_rank_in_infeasible
+        n_below -= n_infeasible
 
     # Third, we calculate the domination rank for trials with no penalty information.
-    ranks, _ = _calculate_nondomination_rank(
-        objective_values[is_penalty_nan], n_below=n_below, base_rank=bottom_rank + 1
+    top_rank_in_penalty_nan = np.max(nondomination_rank[~is_penalty_nan], initial=-1) + 1
+    ranks_in_penalty_nan = _calculate_nondomination_rank(
+        loss_values[is_penalty_nan], n_below=n_below
     )
-    nondomination_rank[is_penalty_nan] += 1 + ranks
-
+    nondomination_rank[is_penalty_nan] = ranks_in_penalty_nan + top_rank_in_penalty_nan
+    assert np.all(nondomination_rank != -1), "All the rank must be updated."
     return nondomination_rank
 
 
+def _is_pareto_front_nd(unique_lexsorted_loss_values: np.ndarray) -> np.ndarray:
+    # NOTE(nabenabe0928): I tried the Kung's algorithm below, but it was not really quick.
+    # https://github.com/optuna/optuna/pull/5302#issuecomment-1988665532
+    loss_values = unique_lexsorted_loss_values.copy()
+    n_trials = loss_values.shape[0]
+    on_front = np.zeros(n_trials, dtype=bool)
+    nondominated_indices = np.arange(n_trials)
+    while len(loss_values):
+        nondominated_and_not_top = np.any(loss_values < loss_values[0], axis=1)
+        # NOTE: trials[j] cannot dominate trials[j] for i < j because of lexsort.
+        # Therefore, nondominated_indices[0] is always non-dominated.
+        on_front[nondominated_indices[0]] = True
+        loss_values = loss_values[nondominated_and_not_top]
+        nondominated_indices = nondominated_indices[nondominated_and_not_top]
+
+    return on_front
+
+
+def _is_pareto_front_2d(unique_lexsorted_loss_values: np.ndarray) -> np.ndarray:
+    n_trials = unique_lexsorted_loss_values.shape[0]
+    cummin_value1 = np.minimum.accumulate(unique_lexsorted_loss_values[:, 1])
+    is_value1_min = cummin_value1 == unique_lexsorted_loss_values[:, 1]
+    is_value1_new_min = cummin_value1[1:] < cummin_value1[:-1]
+    on_front = np.ones(n_trials, dtype=bool)
+    on_front[1:] = is_value1_min[1:] & is_value1_new_min
+    return on_front
+
+
+def _is_pareto_front(unique_lexsorted_loss_values: np.ndarray) -> np.ndarray:
+    (n_trials, n_objectives) = unique_lexsorted_loss_values.shape
+    if n_objectives == 1:
+        return unique_lexsorted_loss_values[:, 0] == unique_lexsorted_loss_values[0]
```

Comment:
```suggestion
        res = np.zeros(len(unique_lexsorted_loss_values))
        res[0] = True
        return res
```
When there is only one objective, the pareto front is always only the first element, since the input is unique.

Reply:
```python
import numpy as np


def impl1():
    x = np.random.random(1000)
    return x == x[0]

def impl2():
    res = np.zeros(1000, dtype=bool)
    res[0] = True
    return res

%timeit impl1() 
%timeit impl2()
4.14 µs ± 26.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
183 ns ± 2.78 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Seems like yours is quicker, so I will take yours!

Reply:
```suggestion
        on_front = np.zeros(len(unique_lexsorted_loss_values), dtype=bool)
        on_front[0] = True  # Only the first element is Pareto optimal.
```